x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java

```diff
@@ -265,36 +263,62 @@ public void messageReceived(T request, TransportChannel channel, Task task) thro
                         }
                     }
                     assert filter != null;
-                    final Thread executingThread = Thread.currentThread();
 
                     final AbstractRunnable receiveMessage = getReceiveRunnable(request, channel, task);
-                    CheckedConsumer<Void, Exception> consumer = (x) -> {
-                        final Executor executor;
-                        if (executingThread == Thread.currentThread()) {
-                            // only fork off if we get called on another thread this means we moved to
-                            // an async execution and in this case we need to go back to the thread pool
-                            // that was actually executing it. it's also possible that the
-                            // thread-pool we are supposed to execute on is `SAME` in that case
-                            // the handler is OK with executing on a network thread and we can just continue even if
-                            // we are on another thread due to async operations
-                            executor = threadPool.executor(ThreadPool.Names.SAME);
-                        } else {
-                            executor = threadPool.executor(executorName);
-                        }
-
-                        try {
-                            executor.execute(receiveMessage);
-                        } catch (Exception e) {
-                            receiveMessage.onFailure(e);
-                        }
-
-                    };
-                    ActionListener<Void> filterListener = ActionListener.wrap(consumer, receiveMessage::onFailure);
+                    final ActionListener<Void> filterListener;
+                    if (ThreadPool.Names.SAME.equals(executorName)) {
+                        filterListener = new AbstractFilterListener(receiveMessage) {
+                            @Override
+                            public void onResponse(Void unused) {
+                                receiveMessage.run();
+                            }
+                        };
+                    } else {
+                        final Thread executingThread = Thread.currentThread();
+                        filterListener = new AbstractFilterListener(receiveMessage) {
+                            @Override
+                            public void onResponse(Void unused) {
+                                if (executingThread == Thread.currentThread()) {
```

Comment:
Is it possible to change this to
```
if (ThreadPool.Names.SAME.equals(executorName) || executingThread == Thread.currentThread()) { ... }
```
so that we can save a few more lines by getting rid of the outer if/else?

Reply:
My thinking was: I figured it was kind of nice to not retain the current `Thread` if we don't have to, that's why I special cased this for `SAME` (we do similar special casing in `InboundHandler` as well).